ddd-koans/Value Object/Extra Task - Sealed Class & Functional Approach/src/Main.kt

```diff
@@ -0,0 +1,42 @@
+sealed class TimeSignature {
+    companion object {
+        private val timeSignatureRegex = "^(\\d{1,2})/(\\d{1,2})$".toRegex()
+
+        fun create(numerator: Int, denominator: Int): TimeSignature =
+            runCatching { ValidTimeSignature(Numerator(numerator), Denominator(denominator)) }
+                .getOrDefault(InvalidTimeSignature)
+
+        fun of(timeSignature: String): TimeSignature =
+            timeSignatureRegex.matchEntire(timeSignature)
+                ?.groupValues
+                ?.takeIf { it.size == 3 }
+                ?.let { group -> group[1] to group[2] }
+                ?.let { (numerator, denominator) -> create(numerator.toInt(), denominator.toInt()) }
+                ?: InvalidTimeSignature
+    }
+}
+
+object InvalidTimeSignature : TimeSignature()
+
+data class ValidTimeSignature(private val numerator: Numerator, private val denominator: Denominator) : TimeSignature()
+
+data class Numerator(private val value: Int) {
+    init {
+        require(value in 1..32)
+    }
+}
+
+data class Denominator(private val value: Int) {
+    init {
+        require(value in 1..32 && value.isPowerOfTwo())
+    }
+}
+
+fun main() {
+    val timeSignaturesToBe = listOf(Pair(1, 1), Pair(1, 2), Pair(3,4), Pair(4, 3))
+
+    //val timeSignatures = timeSignaturesToBe.map { (n, d) -> ValidTimeSignature(Numerator(n), Denominator(d)) }
```

Comment:
When this is uncommented that it will fail as well because `Pair(4, 3)` is not a valid time signature. Moreover, I'm not convinced that using such comment will illustrate the purpose deeply.

I would rather suggest having 2 sets of timeSigniatrues, valid and invalid, to easily grasp which one are invalid. Then merge them together and trying to create a time signature.

Sth like:

```kotlin
    val validTimeSignatures = setOf("1/2", "3/4", "4/4")
    val invalidTimeSignatures = setOf("1/3", "5/7")
    val timeSignatureToCreate = validTimeSignatures + invalidTimeSignatures
    
    timeSignatureToCreate
        .map { input -> TimeSignature.of(input) }
        .forEach(::println)

```

I propose to use `of` as IMO it's easier to understand input values and it also references to previous `static-factory` task